Tweak development tips for Ansible operator (#3236)
operator-framework/operator-sdk: website/content/en/docs/ansible/development-tips.md

This document provides some useful information and tips for a developer
creating an operator powered by Ansible.

## Getting started with the Kubernetes Collection for Ansible

Since we are interested in using Ansible for the lifecycle management of our
application on Kubernetes, it is beneficial for a developer to get a good
grasp of the [Kubernetes Collection for Ansible](https://galaxy.ansible.com/community/kubernetes).
This Ansible collection allows a developer to either leverage their existing
Kubernetes resource files (written in YAML) or express the lifecycle
management in native Ansible. One of the biggest benefits of using Ansible in
conjunction with existing Kubernetes resource files is the ability to use
Jinja templating so that you can customize deployments with the simplicity of
a few variables in Ansible.

The easiest way to get started is to install the collection on your local
machine and test them using a playbook.

### Installing the Kubernetes Collection for Ansible

To install the Kubernetes Collection, one must first install Ansible 2.9+.

> For example, on Fedora/Centos:

Reviewer: I understand that it is about https://docs.ansible.com/ansible/latest/modules/k8s_module.html. Am I right? Could we also add the links and clarify that the following steps is specifically about the collection k8s?

Author: No. We have switched over to the [Kubernetes Collection for Ansible](https://galaxy.ansible.com/community/kubernetes) which provides other modules in addition to `k8s_module`. This new link is added in the revised doc.

Reviewer: Can we please add the info and the link for it?

```bash
$ sudo dnf install ansible
```

In addition to Ansible, a user must install the [OpenShift Restclient

> Python](https://github.com/openshift/openshift-restclient-python) package:

Reviewer: Please, revert it. The links should be added in the bottom.

Author: Okay, I was thinking that maybe the target is only referenced once, we can just put it in place.
For people who work on this doc, scrolling to the bottom to find out the target is a bit inconvenient, unless the target is mentioned several times in the doc.

But anyways, I'll respect the team's convention on this.

```bash
$ pip3 install openshift
```

Finally, install the Kubernetes Collection from ansible-galaxy:

```bash
$ ansible-galaxy collection install community.kubernetes
```

Alternatively, if you've already initialized your operator, you may have a
`requirements.yml` file at the top level of your project. This file specifies
Ansible dependencies that need to be installed for your operator to function.
By default it will install the `community.kubernetes` collection as well as
the `operator_sdk.util` collection, which provides modules and plugins for
operator-specific operations.

To install the dependent modules from this file, run:

```bash
$ ansible-galaxy collection install -r requirements.yml
```

### Testing the Kubernetes Collection locally

Sometimes it is beneficial for a developer to run the Ansible code from their
local machine as opposed to running/rebuilding the operator each time. To do
this, initialize a new project:

```none

> $ operator-sdk new --type ansible --kind Foo --api-version foo.example.com/v1alpha1 foo-operator

Reviewer: Could we replace here for the Memcached one?

Author: See above comment.
If you feel strongly about switching to memcached example, I can revise as well.

Reviewer: See the quick start: https://github.com/operator-framework/operator-sdk/blob/master/website/content/en/docs/ansible/quickstart.md

My suggestion here is: 

```suggestion
$ operator-sdk new memcached-operator --api-version=cache.example.com/v1alpha1 --kind=Memcached --type=ansible
```

Which means explain how the collections work based on in the example that has been provided in the docs already.

Reviewer: I think that would be nice we keep using the same example of the quick start. however, the Memcached is not used now and it should not be a block to move forward with your improvements.

INFO[0000] Creating new Ansible operator 'foo-operator'. 
INFO[0000] Created deploy/service_account.yaml
INFO[0000] Created deploy/role.yaml
[...]
INFO[0000] Created molecule/templates/operator.yaml.j2
INFO[0000] Generated CustomResourceDefinition manifests.
INFO[0000] Project creation complete.

$ cd foo-operator
$ ansible-galaxy collection install -r requirements.yml
```

Modify `roles/foo/tasks/main.yml` with desired Ansible logic. For this example
we will create and delete a ConfigMap based on the value of a variable named
`state`:

> 

Reviewer: Why you change the example to use a ConfigMap instead of the namespace?

TASK [Gathering Facts] *********************************************************************
ok: [localhost]

Task [foo : set example-coonfig configmap to present]
changed: [localhost]

PLAY RECAP *********************************************************************************
localhost                  : ok=2    changed=1    unreachable=0    failed=0

```

Check that the ConfigMap was created:

```none
$ kubectl get configmaps
NAME                    STATUS    AGE

> example-config          Active    3s

Reviewer: all examples are based in the memcached as the sample; https://github.com/operator-framework/operator-sdk-samples/tree/master/ansible/memcached-operator

Can we please use it?

Author: Yes, that is an option. However, my feeling is that for this doc we don't want to force the users to clone a different repo (`git clone https://github.com/operator-framework/operator-sdk-samples.git`) to understand the tips.

The previous doc was using namespace creation as a simple example, which is good. I'm changing it to configmap because namespace is a cluster scoped resource which will fail the in-cluster testing, although everything works fine when running `operator-sdk run local`.

Reviewer: My suggestion was NOT to use the sample. i do not think that we should link it.
just use the Memcached as an example. See the quick start, it will give the steps to create the Memcached sample. IMO we should stick with it instead of creating here `example-config `

Reviewer: I will mark as solved because we have a comment on top about that already.

foo-operator       1         1         1            1           1m
```

Now you can create a Custom Resource instance of Foo with variable `state`
default to `present`:

```bash
$ kubectl create -f deploy/crds/foo.example.com_v1alpha1_foo_cr.yaml
```

Check that ConfigMap `example-config` was created:

```none
$ kubectl get configmaps
NAME                    STATUS    AGE
example-config          Active    3s
```

<!-- TODO(Qiming): consolidate this subsection with quickstart -->
#### Viewing the Ansible logs

In order to see the logs from a particular operator you can run:

```bash
kubectl logs deployment/foo-operator
```

The logs contain the information about the Ansible run and are useful for
debugging your Ansible tasks. Note that the logs may contain much more
detailed information about the Ansible Operator's internals and its
interactions with Kubernetes as well.

Also, you can set the environment variable `ANSIBLE_DEBUG_LOGS` to `True` to
check the full Ansible result in the logs in order to be able to debug it.

**Example**

In the `deploy/operator.yaml`:

```yaml
spec:
  template:
    spec:
      containers:
        - name: foo-operator

>           env:

Reviewer: ```suggestion
          ...
          env:
```

Just to make clear that has more code between the name and the env.

Author: That is doable, but I have already marked this code snippet as YAML. Adding "..." may break the syntax, I'm afraid.
This snippet is obviously not a complete one...

Anyway, if you insist, I can send another commit.

Reviewer: It is the default standard that we work with to illustrate has more code and do not lead misunderstandings. you can check the other docs. Could you please just commit this suggestion? Then, for me it would be ok but before merge requires more one approval which means a second reviewer too.

Author: see 09c91b7

    type: Running
```

An Ansible Operator also allows you to supply custom status values with the
`k8s_status` Ansible module, which is included in

> [operator_sdk.util](https://galaxy.ansible.com/operator_sdk/util) collection.

Reviewer: add the link in the bottom.

Author: Done.

You can update the `status` from within Ansible with any key/value pairs as
desired. If you want to track the CR status manually from your application,

> you can update the `watches.yaml` file with `manageStatus`, as shown below:

Reviewer: Could you revert to the previous description which is more accurate?

Author: Actually I didn't add or remove any text. A summary of the changes made to the paragraph:

- Changed "Ansible Operator" to "An Ansible Operator" to make it generic and grammatically correct.
- Simplified "you as the developer" to "you" because we are in a doc called development tips after all.
- Dropped "By default, Ansible Operator will always include the generic Ansible run output as shown above." because we already said that in previous paragraph.

This only missing piece (after rereading this several times, :)) is `If you would prefer your application *not* update the status with Ansible output`, which I'll add back in next commit.

```

### Ansible Operator Conditions

An Ansible Operator has a set of conditions that are used during reconciliation.
There are mainly two condition types:

* `Running` - the Operator is currently running Ansible for reconciliation or
  if a run has finished with no errors and it is waiting for the next
  reconciliation action. The next action can be triggered by the reconcile

>   period, dependent watches or an update to the Custome Resource.

Reviewer: The : 
```
or
  if a run has finished with no errors and it is waiting for the next
  reconciliation action. The next action can be triggered by the reconcile
  period, dependent watches or an update to the Custome Resource.
```
is not valid for `Running`. 

When the reconciliation finishes it should assume the `Successful` status.

Reviewer: See the code source. The changes made here regards the status are not accurate. 

https://github.com/operator-framework/operator-sdk/blob/ad1d4d1953abf22c336f67198fc23a489c0f4fdb/pkg/ansible/controller/status/utils.go#L22-L30

Author: Well, I don't agree.
The only place the SuccessfulReason is used (referenced) is here - https://github.com/operator-framework/operator-sdk/blob/master/pkg/ansible/controller/reconcile.go#L386-L392. In that context, the condition *type* is still "Running". 

In fact, when we are talking about condition type, rather than the reason why the operator is in a specific condition type, there are only two choices, not three (https://github.com/operator-framework/operator-sdk/blob/master/pkg/ansible/controller/status/types.go#L88-L96)

```go
// ConditionType - type of condition
type ConditionType string

const (
	// RunningConditionType - condition type of running.
	RunningConditionType ConditionType = "Running"
	// FailureConditionType - condition type of failure.
	FailureConditionType ConditionType = "Failure"
)
```

Reviewer: Hi @tenqm,  

If you check the [Memcached Sample](https://github.com/operator-framework/operator-sdk-samples/tree/master/ansible/memcached-operator) when the task finishes the Status will be : 

```
Status:
  Conditions:
    Ansible Result:
      Changed:             0
      Completion:          2020-07-04T09:41:01.05909
      Failures:            0
      Ok:                  2
      Skipped:             0
    Last Transition Time:  2020-07-04T09:40:42Z
    Message:               Awaiting next reconciliation
    Reason:                Successful
    Status:                True
    Type:                  Running
Events:                    <none>
```

See that the type of the conditional is `Running` but the Reason is `Successful` and the doc here is describing the status conditionals reason `Successful` and `Failed`. Because of this, could you please revert your changes and keep it as it was before?

Author: Why revert this? The original text lists three values: `Running`, `Successful` and `Failed`.
The set of condition types:

- `Running`
- `Failure`

The set of condition reasons:

- `Running`: valid when type is `Running`; 
- `Successful`: valid when type is `Running`;
- `Failed`: valid only when type is `Failure`;
- `Unknown`: defined, but not used anywhere, designed for `Failure` condition type.

Reviewer: > Why revert this? The original text lists three values: Running, Successful and Failed.
Because the previous text was describing the `condition reasons` and you are changing the text. 

I agree that it has a hall for improvements. However, WDYT about we revert it and the working on this area in another PR just to improve the explanation over the status conditions and its reasons to make easier the review and etc. Ideally, each pr should have a close/small scope instead with fewer changes in order to ensure the quality and make the process flow smoothly.

Author: When you examine the status of a resource in Kubernetes, you will first look at the condition.type and then the condition.status. The combination of "type" and "status" gives you an idea whether the resource is operational. If you got an negative combination of the "type" and "status", you will check the "message" and "reason" field. In other words, we are not checking the reasons directly.

If we check the condition type definition again, I'm even thinking whether we should remove the `Failure` condition type because it is redundant. A `Failure` type of condition is equivalent to `Running=False`.

| type | status | reason | note |
|------|-------|--------|------|
| Running | True | Running | all good |
| Running | True | Successful | last reconciliation was successful |
| Running | False | Failed | a known error is caught |
| Running | False | Unknown | something bad happened with unknown reason |
| Running | Unknown | ... | ... |
| Failure | True | ... | ... |
| Failure | False | ... | ... |
| Failure | Unknown | ... | ... |

Anyway, I agree we can leave this untouched at the moment.

* `Failure` - if there is any error during reconciliation, the Operator is
  marked as `Failure` with the error message populated. The error message is
  the raw output from the Ansible run. Usually, an intermittent failure is
  automatically resolved when the Operator reruns the reconciliation.

Please check the following sections for debugging an Ansible Operator:

* [View the Ansible logs](../quickstart#view-the-ansible-logs)
* [Additional Ansible debug](../quickstart#additional-ansible-debug)
* [Testing Ansible Operators with Molecule](../testing-guide)

## Extra Variables Sent to Ansible

The Operator treats the content you provide in the `spec` section of the CR
manifest as key-value pairs. It sends them as extra vars  to Ansible. This is
equivalent to [passing extra vars to
`ansible-playbook`](https://docs.ansible.com/ansible/latest/user_guide/playbooks_variables.html#passing-variables-on-the-command-line).
The operator also passes some additional key-value pairs using the `meta` key,

> e.g., the name and the namespace of the CR.

Reviewer: Can you please revert the changes in the explanation and just add the link reference? 
IMO : The previous explanation is more accurate. 
PS.: The links should be added on the bottom. See the other ones.

Author: Moved the link to the bottom of the page.
Would love to understand how we can improve this paragraph for accuracy.

Reviewer: See:

> e.g., the name and the namespace of the CR.

Reviewer: ```suggestion
The Operator treats the content you provide in the `spec` section of the CR
manifest as key-value pairs. It sends them as extra vars to Ansible and are managed by the operator. This is
equivalent to how above extra vars are passed into [passing extra vars to
`ansible-playbook`](https://docs.ansible.com/ansible/latest/user_guide/playbooks_variables.html#passing-variables-on-the-command-line).
The operator also passes along additional variables under the `meta` field. These variable are key-value pairs, such as for example; the CR name and namespace. 
```

Author: > "and are managed by the operator."

What are managed by the operator? 

> The operator also passes along additional variables under the `meta` field. These variable are key-value pairs, such as for example; the CR name and namespace. 

I don't see any where key-value pairs are called "variables". It is confusing. "key-value pairs" contain both the variables and their values, aren't they?  "under the `meta` field" is also a weird expression ...

Reviewer: Hi @tengqm, 

I am not native in English, however, shows that your modifications cange a little the understatement and has grammar issues.  I'd suggest revert and then do a PR with a close scope to address your motivations here. Ideally, we should not do a PR with so many different motivations.

Author: Em... alright. That makes sense.

   },
}
```

The `message` and `newParameter` at the top level are set as extra variables
themselves. The `meta` key contains some metadata relevant to the Custom

> Resource. You can access these metadata  via dot notation in Ansible:

Reviewer: ```suggestion
Resource. The `meta` fields can be accesses via dot notation in Ansible as so:
```

Author: I came from a kubernetes docs community where there is a guide line [use active voice](https://kubernetes.io/docs/contribute/style/style-guide/#use-active-voice).
I'm trying to share some common practices from tech writers, though I am not strongly opinionated on this.

Reviewer: Very interesting. I did not know that it has a definition. I agree we should follow up that. Make senses for me. 
Tks for share I will flag it internally as well.

The `message` and `newParameter` at the top level are set as extra variables
themselves. The `meta` key contains some metadata relevant to the Custom
Resource. You can access these metadata  via dot notation in Ansible:

```yaml
---
- debug:
    msg: "name: {{ meta.name }}, {{ meta.namespace }}"
```

[k8s_ansible_module]:https://docs.ansible.com/ansible/2.6/modules/k8s_module.html

> [openshift_restclient_python]:https://github.com/openshift/openshift-restclient-python

Reviewer: Revert the link. (line 505)
